V14: Adds "Umbraco Flavored Markdown" docs (#6250)
umbraco/UmbracoDocs: 14/umbraco-cms/reference/umbraco-flavored-markdown.md

```
<ufm-label-value alias="bodyText"></ufm-label-value>
```

The internal working of the `ufm-label-value` component would then be able to access the property's value using the [Context API](../extending/backoffice-setup/working-with-data/context-api).

## Available UFM components

As for Umbraco 14.1.0, the following UFM components are available to use.

| Name        | Marker | Example syntax    | Renders component                    |
| ----------- | ------ | ----------------- | ------------------------------------ |
| Label Value | `=`    | `{=bodyText}`     | `<ufm-label-value alias="bodyText">` |
| Localize    | `#`    | `{#general_name}` | `<umb-localize key="general_name">`  |

More UFM components will be available in upcoming Umbraco releases.

If you wish to develop your own custom UFM component, you can use the `ufmComponent` extension type:

```
{
	type: 'ufmComponent',
	alias: 'My.Markdown.CustomComponent',
	name: 'My Custom Markdown Component',
	api: () => import('./components/my-custom.component.js'),
	meta: {
		marker: '%',
	},
}
```

The corresponding JavaScript/TypeScript API would contain a method to render the custom label/markup.

```

> export class MyCustomComponent implements UfmComponentBase {

Reviewer: This would need to extend UmbLitElement, right?

```suggestion
export class MyCustomComponent extends UmbLitElement implements UfmComponentBase {
```

Reviewer: Oh, I see. It's a UFM component. Should we perhaps name the example here in the docs something else?

Author: I'm thinking `MyCustomUfmComponentApi`.

I'd also noticed a typo with the `UfmComponentBase` base class, (it needed the "Umb" prefix). 🤦 
But also, I'm wondering whether `UmbUfmComponentBase` is a suitable name? As it implements the `UmbUfmComponentApi` interface, so maybe the base class should be renamed to `UmbUfmComponentApiBase`? 🤔

Reviewer: I thought we agreed to allow the "Ufm" namespace 😅

# Umbraco Flavored Markdown

Umbraco Flavored Markdown (UFM) is the dialect of Markdown, that is used to support for property descriptions and advanced labels within the Umbraco CMS backoffice. These can be used with Block editors (Block Grid, Block List) and Collection View columns (in Grid and Table views).

{% hint style="info" %}

> **What is Markdown?**
> If you are unfamiliar with Markdown, you can read more about its philosophy and syntax on the Daring Fireball website.
> <https://daringfireball.net/projects/markdown/syntax>

Reviewer 2: ```suggestion
If you are not familiar with Markdown, you can read more about its philosophy and syntax on the [Daring Fireball website](https://daringfireball.net/projects/markdown/syntax).
```

```
export class MyCustomUfmComponentApi implements UmbUfmComponentBase {
	render(token: Tokens.Generic) {
        // You could do further regular expression/text processing here!
		return `<ufm-custom-component text="${token.text}"></ufm-custom-component>`;
	}
}

export { MyCustomUfmComponentApi as api };
```

Using the syntax `{% myCustomText }` would render the markup `<ufm-custom-component text="myCustomText">`. Then inside the `ufm-custom-component` component code, you can perform any logic to render your required markup.

## Post-processing and sanitization

When the markdown has been converted to HTML, the markup will be ran through post-processing sanitization to ensure security and consistency within the backoffice.

As of Umbraco 14, the [DOMPurify library](https://github.com/cure53/DOMPurify) is used to sanitize the markup and prevent XSS attacks.

The sanitizes markup will be...

- Valid HTML
- Anchor links will have their target set to `_blank`
- Only web components that have a prefix of `ufm-`, `umb-` or `uui-` will be allowed to render

## Using UFM in your own components

> If you would like to render UFM within your own web components (within the Umbraco CMS backoffice). You can do so using the `umb-ufm-render` component:

Reviewer 2: ```suggestion
If you would like to render UFM within your own web components in the Umbraco CMS backoffice, you can use the `umb-ufm-render` component:
```
